### PR TITLE
add couchbase vectore store support

### PR DESCRIPTION
**Pull request by prajwal-pai77**
- Add support for Couchbase as a vector store.
- Update Langchain community version to 0.0.43.

### CODE REVIEW

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    badge: string
+    baseClasses: string[]
+    inputs: INodeParams[]
+    credential: INodeParams
+    outputs: INodeOutputsValue[]
+
+    constructor() {
+        this.label = 'Couchbase'
+        this.name = 'Couchbase'
+        this.version = 2.0
```
Comment (HenryHengZJ) [inline packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:27]: version should be `1.0`

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    category: string
+    badge: string
+    baseClasses: string[]
+    inputs: INodeParams[]
+    credential: INodeParams
+    outputs: INodeOutputsValue[]
+
+    constructor() {
+        this.label = 'Couchbase'
+        this.name = 'Couchbase'
```
Comment (HenryHengZJ) [inline packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:26]: name has to follow camelCase: `couchbase`

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        this.type = 'Couchbase'
+        this.icon = 'couchbase.svg'
+        this.category = 'Vector Stores'
+        this.description = `Upsert embedded data and load existing index using Couchbase, a award-winning distributed NoSQL database`
+        this.baseClasses = [this.type, 'VectorStoreRetriever', 'BaseRetriever']
+        this.credential = {
+            label: 'Connect Credential',
+            name: 'credential',
+            type: 'credential',
+            credentialNames: ['CouchbaseApi']
```
Comment (HenryHengZJ) [inline packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:37]: camelCase: `couchbaseApi`

File: packages/components/credentials/CouchbaseApi.credential.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+class CouchbaseApi implements INodeCredential {
+    label: string
+    name: string
+    version: number
+    description: string
+    inputs: INodeParams[]
+
+    constructor() {
+        this.label = 'Couchbase API'
+        this.name = 'CouchbaseApi'
```
Comment (HenryHengZJ) [inline packages/components/credentials/CouchbaseApi.credential.ts:12]: camelCase and version as well

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                if (flattenDocs[i] && flattenDocs[i].pageContent) {
+                    const document = new Document(flattenDocs[i])
+                    finalDocs.push(document)
+                }
+            }
+
+            const couchbaseClient = await Cluster.connect(connectionString, {
+                username: databaseUsername,
+                password: databasePassword,
+                configProfile: 'wanDevelopment'
```
Comment (HenryHengZJ) [inline packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:148]: what is `configProfile`? just trying to understand is it a must to hardcode `wanDevelopment` as the value?

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            embeddingKey: embeddingKey
+        }
+
+        try {
+            if (!textKey || textKey === '') couchbaseConfig.textKey = 'text'
+            if (!embeddingKey || embeddingKey === '') couchbaseConfig.embeddingKey = 'embedding'
+
+            const vectorStore = await CouchbaseVectorStore.initialize(embeddings, couchbaseConfig)
+
+            return resolveVectorStoreOrRetriever(nodeData, vectorStore)
```
Comment (HenryHengZJ) [inline packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:209]: once we have added metadatafilter, we can do:

```
let metadatafilter 
if (couchbaseMetadataFilter) {
            metadatafilter = typeof couchbaseMetadataFilter=== 'object' ? couchbaseMetadataFilter: JSON.parse(couchbaseMetadataFilter)
        }

        const vectorStore = await CouchbaseVectorStore.initialize(embeddings, couchbaseConfig)

        return resolveVectorStoreOrRetriever(nodeData, vectorStore, metadatafilter )
```

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts (hunk 5 above)
Comment (prajwal-pai77) [reply packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:148]: From Couchbase Node.js SDK 4.2 onwards, we have introduced a wanDevelopment Configuration Profile, which provides pre-configured timeout settings suitable for working in high latency environments.
Also this is the recommended way to connect to a cluster as per: https://docs.couchbase.com/nodejs-sdk/current/hello-world/start-using-sdk.html#connect

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+            },
+            {
+                label: 'Embedded Field',
+                name: 'embeddingKey',
+                description: 'Name of the field (column) that contains the Embedding',
+                type: 'string',
+                default: 'embedding',
+                additionalParams: true,
+                optional: true
+            },
```
[the remark this answers is not in the document]
Comment (prajwal-pai77) [reply packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:93]: Yep, thanks added it.

**Comment by prajwal-pai77** [comment]
Hey @HenryHengZJ How can i add couchbase docs in here: https://github.com/FlowiseAI/FlowiseDocs/tree/main/integrations/langchain/vector-stores
What is the process to contribute there?

File: packages/components/nodes/vectorstores/Couchbase/Couchbase.ts (hunk 5 above)
Comment (HenryHengZJ) [reply packages/components/nodes/vectorstores/Couchbase/Couchbase.ts:155]: appreciate the explanation!

**Comment by HenryHengZJ** [comment]
> Hey @HenryHengZJ How can i add couchbase in here: https://github.com/FlowiseAI/FlowiseDocs/tree/main/integrations/langchain/vector-stores What is the process to contribute there?

thanks for updating the docs as well! here's an example [PR](https://github.com/FlowiseAI/FlowiseDocs/pull/48)

**Comment by HenryHengZJ** [comment]
@prajwal-pai77 sorry minor conflict for the pnpm lock file as I've merged in some changes, if you dont mind updating that

**Review by HenryHengZJ — APPROVED** [review]

**Comment by prajwal-pai77** [comment]
Hey @HenryHengZJ  I am having trouble rebasing the pnpm-lock file. But i have managed to push it but it says alot of changes in the pnpm-lock file now, can you please check if this right?
Thanks

**Review by HenryHengZJ — APPROVED** [review]
thanks @prajwal-pai77 !

**Comment by HenryHengZJ** [comment]
> Hey @HenryHengZJ I am having trouble rebasing the pnpm-lock file. But i have managed to push it but it says alot of changes in the pnpm-lock file now, can you please check if this right? Thanks

that's fine, its just the formatting of the lock file that makes the diff large enought to be viewed here

**Comment by HenryHengZJ** [comment]
@prajwal-pai77 there's some problem when trying to create a new docker image:
image

do you have any idea? or had you came across similar issues?

**Comment by prajwal-pai77** [comment]
> @prajwal-pai77 there's some problem when trying to create a new docker image for arm64/amd64: image
> 
> do you have any idea? or had you came across similar issues?

Hey @HenryHengZJ  I see that you have included cmake now in the docker file. Wanted to know if the issue still persists with the docker build. If it does please let me know what the issue.

**Comment by lokesh-couchbase** [comment]
Hey @HenryHengZJ , I went through this and found that the culprit is building the package using Alpine. I checked with the team and as of now we do not support Alpine + Arm64 for Couchbase SDK. 
One of the ways to solve this is if we can shift to using Ubuntu instead of Alpine for the builds as ubuntu uses glibc instead of musl and is much more widely adopted. 
Please let me know if something can be done from your end, and I have expedited the request internally as well for support of musl but it may take some time.

**Comment by HenryHengZJ** [comment]
> Hey @HenryHengZJ , I went through this and found that the culprit is building the package using Alpine. I checked with the team and as of now we do not support Alpine + Arm64 for Couchbase SDK. One of the ways to solve this is if we can shift to using Ubuntu instead of Alpine for the builds as ubuntu uses glibc instead of musl and is much more widely adopted. Please let me know if something can be done from your end, and I have expedited the request internally as well for support of musl but it may take some time.

thanks for figuring out the issue! did some little experiment and using ubuntu as base docker image seems a lil bit too heavy on top of all the dependencies we already have, bringing the total app size from 900mb to around 1.25gb

**Comment by lokesh-couchbase** [comment]
Thanks for testing using Ubuntu. However there is nothing much we can do right now. We still do recommend upgrading to glibc based distro as many of other products might have this as requirement. Even CentOS can be considered.

**Comment by HenryHengZJ** [comment]
> Thanks for testing using Ubuntu. However there is really nothing much we can do right now. We still do recommend upgrading to glibc based distro as many of other products might have this as requirement. Even CentOS can be considered.

yeah, we have a backlog item to work on upgrading the docker to debian based node like bullseye or bookworm, both of which are using glibc. But for now we'll have to temporarily remove the couchbase integration in order to push a new version (some urgent bugfix), until we have the new docker image verified working correctly with all the other integration nodes we have on Flowise

**Pull request merged**